api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -400,13 +400,31 @@ public static Type find(Type type, Predicate<Type> predicate) {
   }
 
   public static boolean isPromotionAllowed(Type from, Type.PrimitiveType to) {
+    return TypeUtil.isPromotionAllowed(from, to, 2);
+  }
+
+  public static boolean isPromotionAllowed(
+      Type from, Type.PrimitiveType to, Integer formatVersion) {
     // Warning! Before changing this function, make sure that the type change doesn't introduce
     // compatibility problems in partitioning.
     if (from.equals(to)) {
       return true;
     }
 
     switch (from.typeId()) {
+      case DATE:
+        if (formatVersion < 3) {
+          return false;
+        } else if (to.typeId() == Type.TypeID.TIMESTAMP) {
+          // Timezone types cannot be promoted.
+          Types.TimestampType toTs = (Types.TimestampType) to;
+          return Types.TimestampType.withoutZone().equals(toTs);
+        } else if (to.typeId() == Type.TypeID.TIMESTAMP_NANO) {
+          // Timezone types cannot be promoted.
+          Types.TimestampNanoType toTs = (Types.TimestampNanoType) to;
+          return Types.TimestampNanoType.withoutZone().equals(toTs);
+        }
+        // fall through
```

Comment:
You did not cover `Date → anything else` case. IMHO You should return false explicitly for any non applicable case.